### PR TITLE
SPU: Implement events channel count, Minor Interrupts fixes

### PR DESCRIPTION
**Pull request by elad335**
The following conditions set the SPU event channel count:

- An event occurs, and the corresponding mask is 1 in the SPU Write Event Mask Channel.
- The SPU Write Event Mask Channel is written with a 1 in a bit position that corresponds to a 1.
- Enabled events are pending after a write of the SPU Write Event Acknowledgment Channel.

The following condition clear the SPU event channel count:
- Reading from SPU Read Event Status Channel.

Interrupt fixes:
Added a missing interrupt invocation check in SPU_WrEventMask in case newly masked events set the events count.

### CODE REVIEW

**Comment by elad335** [comment]
Are there any perf regressions here? otherwise its mergable, I give 2 days for testing.

**Comment by web1018** [comment]
Tested few games using this PR., There were No noticeable performance changes (Drops or Gains) compared to the master version.

- Resistance: Fall of Man
- Resistance 2
- Killzone HD
- Afro Samurai

File: rpcs3/rpcs3qt/register_editor_dialog.cpp [hunk 1]
```diff
@@ -113,13 +114,14 @@ register_editor_dialog::register_editor_dialog(QWidget *parent, u32 _pc, const s
...
 			m_register_combo->addItem("MFC Events Mask", +MFC_EVENTS_MASK);
+			m_register_combo->addItem("MFC Events Count", +MFC_EVENTS_COUNT);
 			m_register_combo->addItem("MFC Tag Mask", +MFC_TAG_MASK);
 			//m_register_combo->addItem("MFC Tag Update", +MFC_TAG_UPD);
 			//m_register_combo->addItem("MFC Atomic Status", +MFC_ATOMIC_STAT);
 			m_register_combo->addItem("SPU SNR1", +SPU_SNR1);
 			m_register_combo->addItem("SPU SNR2", +SPU_SNR2);
-			m_register_combo->addItem("SPU Out Mailbox", +SPU_OUT_MBOX);
-			m_register_combo->addItem("SPU Out-Intr Mailbox", +SPU_OUT_INTR_MBOX);
+			m_register_combo->addItem("Out Mailbox", +SPU_OUT_MBOX);
```
Comment (kd-11) [inline rpcs3/rpcs3qt/register_editor_dialog.cpp:123]: Just, why?

File: rpcs3/rpcs3qt/register_editor_dialog.cpp [hunk 2]
```diff
@@ -187,10 +189,11 @@ void register_editor_dialog::updateRegister(int reg)
...
 			str = fmt::format("%016llx%016llx", spu.gpr[reg_index]._u64[1], spu.gpr[reg_index]._u64[0]);
 		}
-		else if (reg == MFC_PEVENTS) str = fmt::format("%08x", +spu.ch_event_stat);
-		else if (reg == MFC_EVENTS_MASK) str = fmt::format("%08x", +spu.ch_event_mask);
+		else if (reg == MFC_PEVENTS) str = fmt::format("%08x", +spu.ch_events.load().events);
+		else if (reg == MFC_EVENTS_MASK) str = fmt::format("%08x", +spu.ch_events.load().mask);
+		else if (reg == MFC_EVENTS_COUNT) str = fmt::format("%u", +spu.ch_events.load().count);
 		else if (reg == MFC_TAG_MASK) str = fmt::format("%08x", spu.ch_tag_mask);
-		else if (reg == SPU_SRR0) str = fmt::format("%08x", spu.srr0);
+		else if (reg == SPU_SRR0) str = fmt::format("%05x", spu.srr0);
```
Comment (kd-11) [inline rpcs3/rpcs3qt/register_editor_dialog.cpp:196]: Same question here, I would have thought more uniform formatting is easier on the eyes regardless of underlying data structures.

File: rpcs3/Emu/Cell/SPUThread.cpp [hunk 3]
```diff
@@ -872,7 +875,7 @@ std::string spu_thread::dump_regs() const
 		fmt::append(ret, "Reservation Addr: none\n");
 
 	fmt::append(ret, "Atomic Stat: %s\n", ch_atomic_stat); // TODO: use mfc_atomic_status formatting
-	fmt::append(ret, "Interrupts Enabled: %s\n", interrupts_enabled.load());
+	fmt::append(ret, "Interrupts: %s\n", interrupts_enabled ? "Enabeld" : "Disabled");
```
Comment (kd-11) [inline rpcs3/Emu/Cell/SPUThread.cpp:878]: typo

File: rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp [hunk 4]
```diff
@@ -1607,36 +1489,16 @@ void spu_recompiler::RDCH(spu_opcode_t op)
...
-			c->call(imm_ptr(spu_rdch));
-			c->jmp(ret);
-		});
-
-		c->bind(ret);
-		c->movd(x86::xmm0, *addr);
-		c->pslldq(x86::xmm0, 12);
-		c->movdqa(SPU_OFF_128(gpr, op.rt), x86::xmm0);
-		return;
+		break; // TODO
```
Comment (kd-11) [inline rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp:1501]: Multiple TODOs with no real fallback provided here and in WRCH. Doesn't this trash asmjit? What is the plan on this?

**Comment by kd-11** [comment]
I left a review, I like the idea of encapsulation for all the SPU event data into one structure, but there are minor nits. I'm also worried that ASMJIT likely doesn't work anymore yet it has not been relegated to deprecated status, which means we may still have games that run with ASMJIT but not LLVM for some people.

File: rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp (hunk 4 above)
Comment (elad335) [reply rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp:1501]: After the switch there is the main WRCH fallback, unless im missing something.

File: rpcs3/rpcs3qt/register_editor_dialog.cpp (hunk 2 above)
Comment (elad335) [reply rpcs3/rpcs3qt/register_editor_dialog.cpp:196]: srr0's value can span on 5 digits at max (0x3ffff), this is consistent with the SPU debugger as well.

File: rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp (hunk 4 above)
Comment (elad335) [reply rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp:1501]: The new event stuff also added the missing immediate interrupt checks which would be ugly to implement it ASMJIT.

File: rpcs3/rpcs3qt/register_editor_dialog.cpp (hunk 2 above)
Comment (elad335) [reply rpcs3/rpcs3qt/register_editor_dialog.cpp:196]: Well, I can just more error checking instead.

Comment (kd-11) [reply rpcs3/rpcs3qt/register_editor_dialog.cpp:196]: If its already being clamped to 5 digits in other views then its fine, although I personally prefer zero padding to balance out visuals. It's a subjective thing anyway.

**Review by kd-11 — APPROVED** [review]
lgtm

File: rpcs3/rpcs3qt/register_editor_dialog.cpp [hunk 5]
```diff
@@ -318,11 +321,12 @@ void register_editor_dialog::OnOkay(const std::shared_ptr<cpu_thread>& _cpu)
...
 				bool ok = true;
-				if (reg == MFC_PEVENTS && !(reg_value & ~SPU_EVENT_IMPLEMENTED)) spu.ch_event_stat = reg_value;
-				else if (reg == MFC_EVENTS_MASK && !(reg_value & ~SPU_EVENT_IMPLEMENTED)) spu.ch_event_mask = reg_value;
+				if (reg == MFC_PEVENTS && !(reg_value & ~SPU_EVENT_IMPLEMENTED)) spu.ch_events.atomic_op([&](typename spu_thread::ch_events_t& events){ events.events = reg_value; });
+				else if (reg == MFC_EVENTS_MASK && !(reg_value & ~SPU_EVENT_IMPLEMENTED)) spu.ch_events.atomic_op([&](typename spu_thread::ch_events_t& events){ events.mask = reg_value; });
+				else if (reg == MFC_EVENTS_COUNT && reg_value <= 1u) spu.ch_events.atomic_op([&](typename spu_thread::ch_events_t& events){ events.count = reg_value; });
 				else if (reg == MFC_TAG_MASK) spu.ch_tag_mask = reg_value;
-				else if (reg == SPU_SRR0) spu.srr0 = reg_value & 0x3fffc;
-				else if (reg == PC) spu.pc = reg_value & 0x3fffc;
+				else if (reg == SPU_SRR0 && !(value & ~0x3fffc)) spu.srr0 = reg_value;
```
Comment (kd-11) [inline rpcs3/rpcs3qt/register_editor_dialog.cpp:328]: reg_value

**Pull request merged**